mycroft/tts/cache.py

```diff
@@ -0,0 +1,278 @@
+# Copyright 2021 Mycroft AI Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TTS cache maintenance.
+
+There are two types of cache available to a TTS engine.  Both are comprised of
+audio and phoneme files.  TTS engines can use the cache to improve performance
+by not performing inference on sentences in the cache.
+
+The first type of cache is a persistent cache.  The cache is considered
+persistent because the files are stored in a location that is not cleared on
+reboot.  TTS inference on these sentences should only need to occur once.  The
+persistent cache contains commonly spoken sentences.
+
+The second cache type a temporary cache stored in the /tmp directory,
+which is cleared when a device is rebooted.  Sentences are added to this cache
+on the fly every time a TTS engine returns audio for a sentence that is not
+already cached.
+"""
+import base64
+import hashlib
+import json
+import re
+from collections import defaultdict
+from pathlib import Path
+from typing import List, Optional, Set, Tuple
+from urllib import parse
+
+import requests
+
+from mycroft.util.file_utils import (
+    ensure_directory_exists, get_cache_directory
+)
+from mycroft.util.log import LOG
+
+
+def _get_mimic2_audio(sentence: str, url: str) -> Tuple[bytes, str]:
```

Comment:
An idea worth pondering. I struggled with the fact that the call to the Mimic2 API is in two places now.  Ideally this would be better integrated into each TTS engine, but that would have caused API changes that would need to be in a major release.  Maybe a ticket for 21.08?